Nft tutorial using foundry and solmate (#107)
onbjerg/foundry-book: src/tutorials/creating-an-NFT-with-solmate-and-forge.md

# Creating an NFT with solmate and foundry

> This tutorial will run you through the steps create an Opensea compatible NFT and essentially rewrites this [guide](https://docs.opensea.io/docs/getting-started-1), while swapping hardhat  for foundry to deploy, call and test your NFT contract. Furthermore we will replace [OpenZeppelin](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/token/ERC721/ERC721.sol)'s ERC721 library with [Solmate](https://github.com/Rari-Capital/solmate/blob/main/src/tokens/ERC721.sol)'s gas optimised version to implement our NFT contract. A full implementation of this tutorial can be found [here](https://github.com/FredCoen/nft-tutorial).

Reviewer: Some wording and general conciseness

```suggestion
This tutorial walk you through creating an OpenSea compatible NFT with Foundry and [Solmate](https://github.com/Rari-Capital/solmate/blob/main/src/tokens/ERC721.sol). A full implementation of this tutorial can be found [here](https://github.com/FredCoen/nft-tutorial).
```

#### Create project and install dependencies

> Start by setting up a foundry project following the steps outlined in the [Getting started section](https://onbjerg.github.io/foundry-book/getting-started/index.html). Furthermore we will install required dependencies to implement our NFT, namely solmate ERC721 implementation as well as some OpenZeppelin utility libraries. Install the dependencies by running the below from the root of your project:

Reviewer: ```suggestion
Start by setting up a Foundry project following the steps outlined in the [Getting started section](../getting-started). We will also install Solmate for their ERC721 implementation, as well as some OpenZeppelin utility libraries. Install the dependencies by running the following commands from the root of your project:
```

```bash
forge install Rari-Capital/solmate
```

These dependencies will be added as git submodules to your project.

If you have followed the instructions correctly your project should be structured like this:

![Project structure](../images/nft-tutorial/nft-tutorial-project-structure.png)

#### Implement a basic NFT

> Let's rename the boilerplate src/Contract.sol to src/NFT.sol and copy the following code into your renamed file.

Reviewer: ```suggestion
We are then going to rename the boilerplate contract in `src/Contract.sol` to `src/NFT.sol` and replace the code:
```

```
// SPDX-License-Identifier: UNLICENSED
pragma solidity 0.8.10;

import "solmate/tokens/ERC721.sol";

contract NFT is ERC721 {
   uint256 public currentTokenId;

   constructor(
       string memory _name,
       string memory _symbol
   ) ERC721(_name, _symbol) {
   }

   function mintTo(address recipient) public payable returns (uint256) {
       uint256 newItemId = ++currentTokenId;
       _safeMint(recipient, newItemId);
       return newItemId;
   }
}
```

> Let's take a look at this very basic implementation of an NFT. We start by importing to contracts from our git submodules. We import solmate's gas optimised implementation of the ERC721 standard which our NFT contract can inherit from to avoid "reinventing the wheel". Our constructor takes the _name and _symbol arguments for our NFT and passes them on to the constructor of the parent ERC721 implementation. Lastly we implement the mintTo function which allows anyone to mint an NFT from our contract by incrementing the currentTokenId and making use of the _safeMint function of our parent contract.

Reviewer: ```suggestion
Let's take a look at this very basic implementation of an NFT. We start by importing to contracts from our git submodules. We import solmate's gas optimised implementation of the ERC721 standard which our NFT contract will inherit from. Our constructor takes the `_name` and `_symbol` arguments for our NFT and passes them on to the constructor of the parent ERC721 implementation. Lastly we implement the `mintTo` function which allows anyone to mint an NFT. This function increments the `currentTokenId` and makes use of the `_safeMint` function of our parent contract.
```

#### Compile & deploy with forge

> To compile the NFT contract run ```forge build```. By default the compiler output will reside in the out directory. To deploy our compiled contract with forge we have to set environment variables for the RPC endpoint and the private key we want to deploy from.

Reviewer: ```suggestion
To compile the NFT contract run ``forge build``. By default the compiler output will be in the `out` directory. To deploy our compiled contract with Forge we have to set environment variables for the RPC endpoint and the private key we want to use to deploy.
```

Set your environment variables by running:
```bash
export RPC_URL=<Your RPC endpoint>
export PRIVATE_KEY=<Your wallets private key>
```

> Once set you can deploy your NFT with forge running the below command while adding the relevant constructor arguments to the NFT contract:

Reviewer: ```suggestion
Once set, you can deploy your NFT with Forge by running the below command while adding the relevant constructor arguments to the NFT contract:
```

```bash
forge create NFT --rpc-url=$RPC_URL --constructor-args <name> <symbol>
```

> If successfully deployed, you will see the deploying wallet's address, the contract's address as well as the transaction hash printed to your console.

Reviewer: ```suggestion
If successfully deployed, you will see the deploying wallet's address, the contract's address as well as the transaction hash printed to your terminal.
```

#### Minting NFTs from your contract

> Calling functions on your NFT contract is made simple with foundry's cast, a command-line tool for interacting with EVM smart contracts, sending transactions, and getting chain data. Let's have a look at how we can use it to mint NFTs from our NFT contract.

Reviewer: ```suggestion
Calling functions on your NFT contract is made simple with Cast, Foundry's command-line tool for interacting with smart contracts, sending transactions, and getting chain data. Let's have a look at how we can use it to mint NFTs from our NFT contract.
```

Given that you already set your RPC and private key env variables during deployment, mint an NFT from your contract by running:
```bash
cast send --rpc-url=$RPC_URL <contractAddress> "mintTo(address)" <arg>
```

> Well done, you just minted your first NFT from your contract. You can sanity check the owner of the NFT with currentTokenId equal to **0** by running the below ```cast call``` command. The address you provided above should be returned as the owner and printed to your console.

Reviewer: ```suggestion
Well done! You just minted your first NFT from your contract. You can sanity check the owner of the NFT with `currentTokenId` equal to **0** by running the below ``cast call`` command. The address you provided above should be returned as the owner.
```

```bash
cast call --rpc-url=$RPC_URL --private-key=$PRIVATE_KEY <contractAddress> "ownerOf(uint256)" 0
```

#### Extending our NFT contract functionality and testing

> Next, we will extend our NFT by adding metadata to represent the content of our NFTs, as well as set a minting price, a maximum supply and the possibility to withdraw the collected proceeds from minting. To follow along you can replace your current NFT contract with the code snippet below:

Reviewer: ```suggestion
Let's extend our NFT by adding metadata to represent the content of our NFTs, as well as set a minting price, a maximum supply and the possibility to withdraw the collected proceeds from minting. To follow along you can replace your current NFT contract with the code snippet below:
```

    constructor(
        string memory _name,
        string memory _symbol,
        string memory _baseURI
    ) ERC721(_name, _symbol) {
        baseURI = _baseURI;
    }

    function mintTo(address recipient) public payable returns (uint256) {
        require(
            msg.value == MINT_PRICE,
            "Transaction value did not equal the mint price"
        );
        uint256 newTokenId = ++currentTokenId;
        require(newTokenId <= TOTAL_SUPPLY, "Max supply reached");
        _safeMint(recipient, newTokenId);
        return newTokenId;
    }

    function tokenURI(uint256 tokenId)
        public
        view
        virtual
        override
        returns (string memory)
    {
        require(
            ownerOf[tokenId] != address(0),
            "ERC721Metadata: URI query for nonexistent token"
        );
        return
            bytes(baseURI).length > 0
                ? string(abi.encodePacked(baseURI, tokenId.toString()))
                : "";
    }

    /// @dev Overridden in order to make it an onlyOwner function
    function withdrawPayments(address payable payee) public override onlyOwner {
        super.withdrawPayments(payee);
    }
}
```

> Among others we have added metadata that can be queried from any front end application like Opensea by calling the tokenURI method on our Nft contract. **Note**: If you actually want to provide a real url to the constructor at deployment and host the metadata of this NFT contract please follow the steps outlined [here](https://docs.opensea.io/docs/part-3-adding-metadata-and-payments-to-your-contract#intro-to-nft-metadata).

Reviewer: ```suggestion
Among other things, we have added metadata that can be queried from any front-end application like OpenSea, by calling the `tokenURI` method on our NFT contract.

>**Note**: If you want to provide a real URL to the constructor at deployment, and host the metadata of this NFT contract please follow the steps outlined [here](https://docs.opensea.io/docs/part-3-adding-metadata-and-payments-to-your-contract#intro-to-nft-metadata).
```

> Let's test some of this added functionality to make sure it works as intended. Foundry offers an extremely fast EVM native testing framework through forge.

Reviewer: ```suggestion
Let's test some of this added functionality to make sure it works as intended. Foundry offers an extremely fast EVM native testing framework through Forge.
```

> Within your test folder rename the current Contract.t.sol test file to NFT.t.sol. This file will contain all tests regarding the NFT's mintTo method. Next, replace the existing boilerplate code with the below:

Reviewer: ```suggestion
Within your test folder rename the current `Contract.t.sol` test file to `NFT.t.sol`. This file will contain all tests regarding the NFT's `mintTo` method. Next, replace the existing boilerplate code with the below:
```

        nft.mintTo{value: 0.08 ether}(address(1));
        uint256 slotBalance = stdstore
            .target(address(nft))
            .sig(nft.balanceOf.selector)
            .with_key(address(1))
            .find();
        
        uint256 balanceFirstMint = uint256(vm.load(address(nft), bytes32(slotBalance)));
        assertEq(balanceFirstMint, 1);

        nft.mintTo{value: 0.08 ether}(address(1));
        uint256 balanceSecondMint = uint256(vm.load(address(nft), bytes32(slotBalance)));
        assertEq(balanceSecondMint, 2);
    }

    function testSafeContractReceiver() public {
        Receiver receiver = new Receiver();
        nft.mintTo{value: 0.08 ether}(address(receiver));
         uint256 slotBalance = stdstore
            .target(address(nft))
            .sig(nft.balanceOf.selector)
            .with_key(address(receiver))
            .find();

        uint256 balance = uint256(vm.load(address(nft), bytes32(slotBalance)));
        assertEq(balance, 1);
    }
    
    function testFailUnSafeContractReceiver() public {
        vm.etch(address(1), bytes("mock code"));
        nft.mintTo{value: 0.08 ether}(address(1));
    }
}

> 

Reviewer: ```suggestion
```

contract Receiver is ERC721TokenReceiver {
    function onERC721Received(
        address operator,
        address from,
        uint256 id,
        bytes calldata data
    ) external returns (bytes4){
        return this.onERC721Received.selector;
    }
}
```

> The test suite is set up as a contract with a setUp method which runs before every individual test. As you can see forge offers a number of [cheatcodes](https://onbjerg.github.io/foundry-book/forge/cheatcodes.html) to manipulate your state to accommodate to your testing scenario. For example, our testFailMaxSupplyReached test  checks that an attempt to mint fails when the max supply of NFT is reached. Thus the currentTokenId of the NFT contract needs to be set to the max supply by using the store cheat code which allows you to write data to your contracts storage slots. The storage slots you wish to write to can easily be found using the

Reviewer: ```suggestion
The test suite is set up as a contract with a `setUp` method which runs before every individual test.

As you can see, Forge offers a number of [cheatcodes](../forge/cheatcodes) to manipulate state to accomodate your testing scenario.

For example, our `testFailMaxSupplyReached` test checks that an attempt to mint fails when the max supply of NFT is reached. Thus, the `currentTokenId` of the NFT contract needs to be set to the max supply by using the store cheatcode which allows you to write data to your contracts storage slots. The storage slots you wish to write to can easily be found using the
```

> forge-std helper library. You can run the test with the following command:

Reviewer: ```suggestion
[`forge-std`](https://github.com/brockelmore/forge-std/) helper library. You can run the test with the following command:
```

```bash
forge test
```

> If you want to put your forge skills to practice, write tests for the remaining methods of our NFT contract. Feel free to PR them [nft-tutorial]((https://github.com/FredCoen/nft-tutorial) where you will find the full implemenation of this tutorial.

Reviewer: ```suggestion
If you want to put your Forge skills to practice, write tests for the remaining methods of our NFT contract. Feel free to PR them to [nft-tutorial]((https://github.com/FredCoen/nft-tutorial), where you will find the full implemenation of this tutorial.
```

#### Gas reports for your function calls

> Foundry provides a comprehensive gas report about your contracts. For every function called within your tests it returns the min, average, median and max gas cost. To print the gas report simply run:

Reviewer: ```suggestion
Foundry provides comprehensive gas reports about your contracts. For every function called within your tests it, returns the minimum, average, median and max gas cost. To print the gas report simply run:
```

```bash
forge test --gas-report
```

> This comes in handy when looking at various gas optimizations within your contracts. Since we used Solmate gas optimised ERC721 library instead of OpenZeppelin's ERC721 library, let's have a look at the gas savings we made. You can find the NFT implementation using both libraries [here](https://github.com/FredCoen/nft-tutorial). Below are the resulting gas reports when running ```forge test --gas-report``` on that repository. As you can see our implementation using solmate library saves around 500 gas on a successful mint (the max gas cost of the mintTo function calls).

Reviewer: ```suggestion
This comes in handy when looking at various gas optimizations within your contracts.

Let's have a look at the gas savings we made by substituting OpenZeppelin with Solmate for our ERC721 implementation. You can find the NFT implementation using both libraries [here](https://github.com/FredCoen/nft-tutorial). Below are the resulting gas reports when running ``forge test --gas-report`` on that repository.

As you can see, our implementation using Solmate saves around 500 gas on a successful mint (the max gas cost of the `mintTo` function calls).
```

```bash
forge create NFT --rpc-url=$RPC_URL --private-key=$PRIVATE_KEY --constructor-args <name> <symbol>
```

If successfully deployed, you will see the deploying wallet's address, the contract's address as well as the transaction hash printed to your console.

#### Minting NFTs from your contract

Calling functions on your NFT contract is made simple with foundry's cast, a command-line tool for interacting with EVM smart contracts, sending transactions, and getting chain data. Let's have a look at how we can use it to mint NFTs from our NFT contract.

Given that you already set your RPC and private key env variables during deployment, mint an NFT from your contract by running:
```bash
cast send --rpc-url=$RPC_URL --private-key=$PRIVATE_KEY <contractAddress> "mintTo(address)" <arg>
```

Well done, you just minted your first NFT from your contract. You can sanity check the owner of the NFT with currentTokenId equal to **0** by running the below ```cast call``` command. The address you provided above should be returned as the owner and printed to your console.

```bash

> cast call --rpc-url=$RPC_URL --private-key=$PRIVATE_KEY <contractAddress> "ownerOf(uint256)" 0

Reviewer 2: not needed

```suggestion
cast call --rpc-url=$RPC_URL <contractAddress> "ownerOf(uint256)(address)" 0
```

```

#### Extending our NFT contract functionality and testing

Next, we will extend our NFT by adding metadata to represent the content of our NFTs, as well as set a minting price, a maximum supply and the possibility to withdraw the collected proceeds from minting. To follow along you can replace your current NFT contract with the code snippet below:

```
// SPDX-License-Identifier: UNLICENSED
pragma solidity 0.8.10;

import "solmate/tokens/ERC721.sol";

> import "openzeppelin-contracts/contracts/utils/Counters.sol";

Reviewer 2: no counters sir

Let's test some of this added functionality to make sure it works as intended. Foundry offers an extremely fast EVM native testing framework through forge.

Within your test folder rename the current Contract.t.sol test file to NFT.t.sol. This file will contain all tests regarding the NFT's mintTo method. Next, replace the existing boilerplate code with the below:

```
// SPDX-License-Identifier: UNLICENSED
pragma solidity 0.8.10;

import "ds-test/test.sol";
import "forge-std/stdlib.sol";
import "../NFT.sol";
import "./interfaces/HEVM.sol";

contract NFT is DSTest {
   using stdStorage for StdStorage;

   Hevm private vm = Hevm(HEVM_ADDRESS);
   NFT private nft;
   StdStorage private stdstore;

   function setUp() public {
       // Deploy NFT contract
       nft = new NFT("NFT_tutorial", "TUT", "");
   }

   function testFailNoMintPricePaid() public {
       nft.mintTo(address(1));
   }

   function testMintPricePaid() public {
       nft.mintTo{value: 0.08 ether}(address(1));
   }

   function testFailMaxSupplyReached() public {
       uint256 slot = stdstore.target(address(nft)).sig("currentTokenId()").depth(0).find();

>        bytes32 loc = bytes32(abi.encode(slot));

Reviewer 2: Think the `abi.encode` is not needed?

           .with_key(address(1))
           .find();
      
       uint256 balanceFirstMint = uint256(vm.load(address(nft), bytes32(slotBalance)));
       assertEq(balanceFirstMint, 1);

       nft.mintTo{value: 0.08 ether}(address(1));
       uint256 balanceSecondMint = uint256(vm.load(address(nft), bytes32(slotBalance)));
       assertEq(balanceSecondMint, 2);
   }

   function testSafeContractReceiver() public {
       Receiver receiver = new Receiver();
       nft.mintTo{value: 0.08 ether}(address(receiver));
        uint256 slotBalance = stdstore
           .target(address(nft))
           .sig(nft.balanceOf.selector)
           .with_key(address(receiver))
           .find();

       uint256 balance = uint256(vm.load(address(nft), bytes32(slotBalance)));
       assertEq(balance, 1);
   }
  
   function testFailUnSafeContractReceiver() public {
       vm.etch(address(1), bytes("mock code"));
       nft.mintTo{value: 0.08 ether}(address(1));
   }
}

contract Receiver is ERC721TokenReceiver {
   function onERC721Received(
       address operator,
       address from,
       uint256 id,
       bytes calldata data
   ) external returns (bytes4){

>        return bytes4(keccak256("onERC721Received(address,address,uint256,bytes)"));

Reviewer 2: think you can make this `this.onERC721Received.selector`

#### Gas reports for your function calls

Foundry provides a comprehensive gas report about your contracts. For every function called within your tests it returns the min, average, median and max gas cost. To print the gas report simply run:

```bash
forge test --gas-report
```

This comes in handy when looking at various gas optimizations within your contracts. Since we used Solmate gas optimised ERC721 library instead of OpenZeppelin's ERC721 library, let's have a look at the gas savings we made. You can find the NFT implementation using both libraries [here](https://github.com/FredCoen/nft-tutorial). Below are the resulting gas reports when running ```forge test --gas-report``` on that repository. As you can see our implementation using solmate library saves around 1000 gas on a successful mint (the max gas cost of the mintTo function calls).

![Gas report solmate NFT](../images/nft-tutorial/gas-report-solmate-nft.png)

![Gas report OZ NFT](../images/nft-tutorial/gas-report-oz-nft.png)

> That's it, I hope this will give you a good practical basis of how to get started with foundry. In my opinion there is no better way to deeply understand solidity than writing your tests in solidity. You will also experience less context switching between javascript and solidity. Happy coding!

Reviewer 2: Would recommend changing this to use "we" or avoid referring to self altogether (i.e. use third person)
